AWS service guide (#818)
crossplaneio/crossplane: docs/services/services-guide-aws.md

> # Deploying Wordpress in Amazon Web Services (AWS)

Reviewer: This is missing front-matter to include it in the ToC:

```yaml
---
title: Using AWS Services
toc: true
weight: 440
indent: true
---
```

This user guide will walk you through Wordpress application deployment using
Crossplane managed resources and the official Wordpress Docker image.

## Table of Contents

1. [Pre-requisites](#pre-requisites)
1. [Preparation](#preparation)
1. [Set Up an EKS Cluster](#set-up-an-eks-cluster)
1. [Set Up RDS Configurations](#set-up-rds-configurations)
1. [Set Up Crossplane](#set-up-crossplane)
1. [Install Wordpress](#install-wordpress)
1. [Uninstall](#uninstall)
1. [Conclusion and Next Steps](#conclusion-and-next-steps)

## Pre-requisites

These tools are required to complete this guide. They must be installed on your
local machine.

* [AWS CLI][aws-cli]
* [kubectl][install-kubectl]
* [Helm][using-helm], minimum version `v2.10.0+`.
* [jq][jq-docs] - commandline JSON processor `v1.5+`

## Preparation

This guide assumes that you have setup and configured the AWS CLI.

*Note: these session variables are used throughout this guide. You may use the
values below or create your own.*

### Set Up an EKS Cluster

> We will create an EKS cluster, follwoing the steps provided in [AWS documentation] [aws-create-eks].

Reviewer 2: The `[AWS documentation]` link is dangling

```bash
aws cloudformation describe-stacks --output json --stack-name ${EKS_STACK_NAME} --region $REGION | jq -r '.Stacks[0].StackStatus'
```

Once the output of the above command is `CREATE_COMPLETE`, the stack creation is completed, and you can retrieve some of the properties of the created resources. These properties will later be consumed in other resources.

```bash
VPC_ID=$(aws cloudformation describe-stacks --output json --stack-name ${EKS_STACK_NAME} --region ${REGION} | jq -r '.Stacks[0].Outputs[]|select(.OutputKey=="VpcId").OutputValue')

# comma separated list of Subnet IDs
SUBNET_IDS=$(aws cloudformation describe-stacks --output json --stack-name ${EKS_STACK_NAME} --region ${REGION} | jq -r '.Stacks[0].Outputs[]|select(.OutputKey=="SubnetIds").OutputValue')

# the ID of security group that later will be used for the EKS cluster
EKS_SECURITY_GROUP=$(aws cloudformation describe-stacks --output json --stack-name ${EKS_STACK_NAME} --region ${REGION} | jq -r '.Stacks[0].Outputs[]|select(.OutputKey=="SecurityGroups").OutputValue')
```

#### Create an IAM Role for the EKS cluster

> For EKS cluster to be able to access different resources, it needs to be given the required permissions through an **IAM Role**. In this section we create a role and assign the required policies. Later we will make EKS to assume this role.

Reviewer 2: ```suggestion
For the EKS cluster to be able to access different resources, it needs to be given the required permissions through an **IAM Role**. In this section we create a role and assign the required policies. Later we will make EKS assume this role.
```

```bash
EKS_ROLE_NAME=crossplane-example-eks-role

# the policy that determines what principal can assume this role
ASSUME_POLICY='{
  "Version": "2012-10-17",
  "Statement": {
    "Effect": "Allow", 
    "Principal": {
      "Service": "eks.amazonaws.com"
    },
    "Action": "sts:AssumeRole"
  }
}'

# Create a Role that can be assumed by EKS
aws iam create-role \
  --role-name "${EKS_ROLE_NAME}" \
  --region "${REGION}" \
  --assume-role-policy-document "${ASSUME_POLICY}"
```

The output should be the created role in JSON. Next, you'll attach the required policies to this role:

```bash
# attach the required policies
aws iam attach-role-policy --role-name "${EKS_ROLE_NAME}" --policy-arn arn:aws:iam::aws:policy/AmazonEKSClusterPolicy > /dev/null
aws iam attach-role-policy --role-name "${EKS_ROLE_NAME}" --policy-arn arn:aws:iam::aws:policy/AmazonEKSServicePolicy > /dev/null
```

> Now lets retrieve the **arn** of this role and store it in a variable. We later assign this to the EKS.

Reviewer 2: ```suggestion
Now lets retrieve the **ARN** of this role and store it in a variable. We later assign this to the EKS.
```

```bash
EKS_ROLE_ARN=$(aws iam get-role --output json --role-name "${EKS_ROLE_NAME}" | jq -r .Role.Arn)
```

#### Creating the EKS cluster

At this step we create an EKS cluster:

```bash
CLUSTER_NAME=crossplane-example-cluster
K8S_VERSION=1.14

aws eks create-cluster \
  --region "${REGION}" \
  --name "${CLUSTER_NAME}" \
  --kubernetes-version "${K8S_VERSION}" \
  --role-arn "${EKS_ROLE_ARN}" \
  --resources-vpc-config subnetIds="${SUBNET_IDS}",securityGroupIds="${EKS_SECURITY_GROUP}"
```

> At this point, an EKS cluster should be started provisioning, which could take take up to 15 minutes. You can check the status of the cluster by running:

Reviewer 2: ```suggestion
At this point, an EKS cluster should have started provisioning, which could take take up to 15 minutes. You can check the status of the cluster by running:
```

```bash
aws eks describe-cluster --name "${CLUSTER_NAME}" --region "${REGION}" | jq -r .cluster.status
```

Once the provisioning is completed, the above command will return `ACTIVE`.

> #### Configuring `kubectl` to communicatie with the EKS 

Reviewer 2: ```suggestion
#### Configuring `kubectl` to communicate with the EKS cluster
```

> Once the cluster is created and is `ACTIVE`, we configure the `kubectl` to target this cluster:

Reviewer 2: ```suggestion
Once the cluster is created and is `ACTIVE`, we configure `kubectl` to target it:
```

```bash
# this environment variable tells kubectl what config file to use
# its value is an arbitrary name, which will be used to name
# the eks cluster configuration file
export KUBECONFIG=~/.kube/eks-config

# this command will populate the eks k8s config file
aws eks update-kubeconfig \
  --name  "${CLUSTER_NAME}"\
  --region "${REGION}"\
  --kubeconfig "${KUBECONFIG}"
```

The output will look like:

>```bash
>Added new context arn:aws:eks:eu-west-1:123456789012:cluster/crossplane-example-cluster to /path/to/.kube/eks-config
>```

At this point, `kubectl` should be configurd to talk to the EKS cluster. To verify this run:

```bash
kubectl cluster-info
```

Which should produce something like:

>```bash
>Kubernetes master is running at https://12E34567898A607F40B3C2FDDF42DC5.sk1.eu-west-1.eks.amazonaws.com
>```

#### Creating Worker Nodes for the EKS cluster

> The worker noes of an EKS cluster are not managed by the cluster itself. Instead, a set of EC2 instances, along with other resources and configurations are needed to be provisioned. In this section, we will create another Cloud Formation stack to setup a worker node configuration, which is described in [EKS official documentation][sample-workernodes-stack].

Reviewer 2: ```suggestion
The worker nodes of an EKS cluster are not managed by the cluster itself. Instead, a set of EC2 instances, along with other resources and configurations need to be provisioned. In this section, we will create another Cloud Formation stack to setup a worker node configuration, which is described in [EKS official documentation][sample-workernodes-stack].
```

# which we do not need and can ignore
aws ec2 create-key-pair --key-name "${KEY_PAIR}" --region="${REGION}" > /dev/null

# create the workers stack
aws cloudformation create-stack \
  --template-url="${WORKERS_STACK_URL}" \
  --region "${REGION}" \
  --stack-name "${WORKERS_STACK_NAME}" \
   --capabilities CAPABILITY_IAM \
  --parameters \
    ParameterKey=ClusterName,ParameterValue="${CLUSTER_NAME}" \
    ParameterKey=ClusterControlPlaneSecurityGroup,ParameterValue="${EKS_SECURITY_GROUP}" \
    ParameterKey=NodeGroupName,ParameterValue="crossplane-eks-nodes" \
    ParameterKey=NodeAutoScalingGroupMinSize,ParameterValue=1 \
    ParameterKey=NodeAutoScalingGroupDesiredCapacity,ParameterValue=1 \
    ParameterKey=NodeInstanceType,ParameterValue="m3.medium" \
    ParameterKey=NodeImageId,ParameterValue="${NODE_IMAGE_ID}" \
    ParameterKey=KeyName,ParameterValue="${KEY_PAIR}" \
    ParameterKey=VpcId,ParameterValue="${VPC_ID}" \
    ParameterKey=Subnets,ParameterValue="${SUBNET_IDS//,/\,}"
```

The output will look like:

>```bash
>{
>    "StackId": "arn:aws:cloudformation:eu-west-1:123456789012:stack/aws-serivce-example-stack 5730d720-d9d9-11e9-8662-029e8e947a9c"
>}
>```

> Similar to VPC stack, creating the workers stack continutes in the background and could take a few minutes to complete. You can check its status by running:

Reviewer 2: ```suggestion
Similar to the VPC stack, creating the workers stack continues in the background and could take a few minutes to complete. You can check its status by running:
```

```bash
NODE_INSTANCE_ROLE=$(aws cloudformation describe-stacks --output json --stack-name ${WORKERS_STACK_NAME} --region ${REGION} | jq -r '.Stacks[0].Outputs[]|select(.OutputKey=="NodeInstanceRole").OutputValue')

cat > aws-auth-cm.yaml <<EOF
apiVersion: v1
kind: ConfigMap
metadata:
  name: aws-auth
  namespace: kube-system
data:
  mapRoles: |
    - rolearn: ${NODE_INSTANCE_ROLE}
      username: system:node:{{EC2PrivateDNSName}}
      groups:
        - system:bootstrappers
        - system:nodes
EOF
```

Then, we will apply this config map to the EKS cluster:

```bash
kubectl apply -f aws-auth-cm.yaml
```

This should print the following output:

>```bash
> configmap/aws-auth created
>```

Now, you can monitor that worker nodes (in this case only a single node) join the cluster:

```bash
kubectl get nodes
```

>```bash
> ip-192-168-104-194.eu-west-1.compute.internal   NotReady   <none>   8s    v1.14.6-eks-5047ed
>```

After a few minutes, the `NotReady` status above should become `Ready`.

Congratulations! You have successfully setup and configured your EKS cluster!

### Set Up RDS Configurations

> In AWS an RDS database instance, will be provisioned to satisfy WordPress application's `MySQLInstanceClass` claim. In order to make an RDS instance accessible by the EKS cluster:

Reviewer 2: ```suggestion
In AWS an RDS database instance will be provisioned to satisfy WordPress application's `MySQLInstanceClass` claim. In order to make an RDS instance accessible by the EKS cluster:
```

> 1. A security group should be created and assigned to the RDS, so certain traffic from the EKS is allowed

Reviewer 2: ```suggestion
1. A security group should be created and assigned to the RDS, so certain traffic from the EKS cluster is allowed
```

```bash

# an arbitrary name for the security group
RDS_SG_NAME="crossplane-example-rds-sg"

# Generate the Security Group and add MySQL ingress to it
aws ec2 create-security-group \
  --vpc-id="${VPC_ID}" \
  --region="${REGION}" \
  --group-name="${RDS_SG_NAME}" \
  --description="open mysql access for crossplane-example cluster" 

# retrieve the ID for this security group
RDS_SECURITY_GROUP_ID=$(aws ec2 describe-security-groups --filter Name=group-name,Values="${RDS_SG_NAME}" --region="${REGION}" --output=text --query="SecurityGroups[0].GroupId")
```

> After creating the security group, we add a rule to allow traffic at `MySQL` port

Reviewer 2: ```suggestion
After creating the security group, we add a rule to allow traffic on `MySQL`'s port
```

```bash
aws ec2 authorize-security-group-ingress \
  --group-id="${RDS_SECURITY_GROUP_ID}" \
  --protocol=tcp \
  --port=3306 \
  --region="${REGION}" \
  --cidr=0.0.0.0/0  > /dev/null
```

1. A **DB Subnet Goup** needs to be created, so that the RDS instance is associated with different availability zones

```bash

DB_SUBNET_GROUP_NAME=crossplane-example-dbsubnetgroup

# convert subnets to a white space separated list
SUBNETS_LIST="${SUBNET_IDS//,/ }"

aws rds create-db-subnet-group \
  --region="${REGION}" \
  --db-subnet-group-name="${DB_SUBNET_GROUP_NAME}" \
  --db-subnet-group-description="crossplane-example db subnet group" \
  --subnet-ids $SUBNETS_LIST > /dev/null

```

> These resources later will be used to create cloud-specitic MySQL resources.

Reviewer 2: ```suggestion
These resources will later be used to create cloud-specific MySQL resources.
```

### Set Up Crossplane

Using the newly provisioned cluster:

> 1. Install Crossplane from alpha channel. (See the [Crossplane Installation

Reviewer 2: ```suggestion
1. Install Crossplane from the alpha channel. (See the [Crossplane Installation
```

```

2. Install the AWS stack into Crossplane. (See the [AWS stack
   section][aws-stack-install] of the install guide for more information.)

```yaml
cat > stack-aws.yaml <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: aws
---
apiVersion: stacks.crossplane.io/v1alpha1
kind: StackRequest
metadata:
  name: stack-aws
  namespace: crossplane-system
spec:
  package: "crossplane/stack-aws:master"
EOF

kubectl apply -f stack-aws.yaml
```

3. Obtain AWS credentials. (See the [Cloud Provider Credentials][cloud-creds]
   docs for more information.)

#### Infrastructure Namespaces

Kubernetes namespaces allow for separation of environments within your cluster.
You may choose to use namespaces to group resources by team, application, or any
other logical distinction. For this guide, we will create a namespace called
`app-project1-dev`, which we will use to group our AWS infrastructure
components.

* Define a `Namespace` in `aws-infra-dev-namespace.yaml` and create it:
```yaml
cat > aws-infra-dev-namespace.yaml <<EOF
---
apiVersion: v1
kind: Namespace
metadata:
  name: aws-infra-dev
EOF

kubectl apply -f aws-infra-dev-namespace.yam
```

* You should see the following output:

> > namespace/aws-infra-dev.yaml created

Reviewer 2: ```suggestion
> namespace/aws-infra-dev created
```

(I think?)

#### AWS Provider

> It is essential to make sure that the AWS user credentials is configured

Reviewer 2: ```suggestion
It is essential to make sure that the AWS user credentials are configured
```

in Crossplane as a provider. Please follow the steps [provider guide][aws-provider-guide]
for more information.

#### Cloud-Specific Resource Classes

Cloud-specific resource classes are used to define a reusable configuration for

> a specific managed service. Wordpress requires a MySQL database, which can be

Reviewer 2: ```suggestion
a specific managed resource. Wordpress requires a MySQL database, which can be
```

I think we more commonly say managed resource (meaning the actual `RDSInstance`), but managed service makes sense here too if you mean for our RDS support in general.

To delete all created resources, but leave Crossplane and the AWS stack
running, execute the following commands:
```bash
kubectl delete -f mysql-claim.yaml
kubectl delete -f mysql-standard.yaml
kubectl delete -f aws-mysql-standard.yaml
kubectl delete -f app-project1-dev-namespace.yaml
kubectl delete -f aws-infra-dev-namespace.yaml
```

## Conclusion and Next Steps

In this guide we:

* Setup an EKS Cluster using the AWS CLI
* Configured RDS to communicate with EKS
* Installed Crossplane from alpha channel
* Installed the AWS stack
* Created an infrastructure (`aws-infra-dev`) and application
  (`app-project1-dev`) namespace
* Setup an AWS `Provider` with our account
* Created a `RDSInstanceClass` in the ` with configuration using RDS configuration earlier
* Created a `MySQLInstanceClass` that specified the `RDSInstanceClass` as
  `mysql-standard` in the `app-project1-dev` namespace
* Created a `MySQLInstance` claim in the `app-project1-dev1` namespace that
  referenced `mysql-standard`
* Created a `Deployment` and `Service` to run Wordpress on our EKS Cluster and
  assign an external IP address to it

If you would like to try out a similar workflow using a different cloud

> provider, take a look at the other [services guides][serives]. If you would like

Reviewer 2: ```suggestion
provider, take a look at the other [services guides][services]. If you would like
```
